packages/dma-library/src/utils/morpho/reallocate.ts

```diff
@@ -0,0 +1,330 @@
+  }
+  oracle: {
+    address: string
+  }
+  irmAddress: string
+  lltv: number
+}
+
+/**
+ * @notice Query market data using a unique key and chain ID.
+ * @param uniqueKey The unique key for the market.
+ * @param chainId The ID of the blockchain network.
+ * @return The market data.
+ */
+export const queryMarketData = async (
+  uniqueKey: string,
+  chainId: number,
+): Promise<MarketByUniqueKeyResponse> => {
+  const query = queries.query2
+  const data: any = await fetchAPI(query, { uniqueKey, chainId })
+  return data?.data?.marketByUniqueKey || {}
+}
+
+/**
+ * @notice Extract data from market data for withdrawals and supply market parameters.
+ * @param marketData The market data object.
+ * @return An object containing withdrawals grouped by vault and supply market parameters.
+ */
+const extractDataForReallocation = (
+  marketData: MarketByUniqueKeyResponse,
+  liquidity: BigNumber,
+) => {
+  const withdrawalsPerVault: { [vaultAddress: string]: Withdrawal[] } = {}
+  const availableLiquidity = new BigNumber(marketData.state.liquidityAssets)
+  // Subtract the liquidity available in the market from the total liquidity required
+  let remainingLiquidity = liquidity.minus(availableLiquidity)
+  // Add a buffer factor to the remaining liquidity ( account for constatnt interest accrual, stale calldata etc )
+  remainingLiquidity = remainingLiquidity.times(BUFFER_FACTOR)
+
+  // First, group and sum assets by vault
+  const vaultTotalAssets = marketData.publicAllocatorSharedLiquidity.reduce(
+    (acc: { [key: string]: BigNumber }, item: any) => {
+      const vaultAddress = item.vault.address
+      acc[vaultAddress] = (acc[vaultAddress] || new BigNumber(0)).plus(item.assets)
+      return acc
+    },
+    {},
+  )
+
+  // Sort vaults by total assets (descending)
+  const sortedVaults = Object.entries(vaultTotalAssets).sort(([, a], [, b]) =>
+    b.minus(a).toNumber(),
+  )
+
+  // Process each vault's allocations
+  for (const [vaultAddress] of sortedVaults) {
+    if (remainingLiquidity.lte(0)) break
+
+    const vaultAllocations = marketData.publicAllocatorSharedLiquidity.filter(
+      (item: any) => item.vault.address === vaultAddress,
+    )
+
+    for (const item of vaultAllocations) {
+      const itemAmount = new BigNumber(item.assets)
+
+      // Skip if we've already collected enough liquidity
+      if (remainingLiquidity.lte(0)) break
+
+      // Calculate how much we can take from this allocation
+      const amountToTake = BigNumber.minimum(itemAmount, remainingLiquidity)
+      remainingLiquidity = remainingLiquidity.minus(amountToTake)
+
+      const withdrawal: Withdrawal = {
+        marketParams: {
+          loanToken: item.allocationMarket.loanAsset.address,
+          collateralToken: item.allocationMarket.collateralAsset.address,
+          oracle: item.allocationMarket.oracle.address,
+          irm: item.allocationMarket.irmAddress,
+          lltv: item.allocationMarket.lltv.toString(),
+        },
```

Comment:
In case the market is what we call the "idle market", the api is wrongly returning a null value instead of the zero address

For instance I jumped on this once:
  allocationMarket: {
    uniqueKey: '0x38c846197ac32a752a60c25d4536ebb0c3920c532e9a859c38c91efb7b8c2abb',
    loanAsset: { address: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913' },
    collateralAsset: null,
    irmAddress: '0x0000000000000000000000000000000000000000',
    oracle: null,
    lltv: 0
  }
  
```suggestion
      const withdrawal: Withdrawal = {
        marketParams: {
          loanToken: item.allocationMarket.loanAsset.address,
          collateralToken: item.allocationMarket.collateralAsset?.address ||
            "0x0000000000000000000000000000000000000000",
          oracle: item.allocationMarket.oracle?.address ||
            "0x0000000000000000000000000000000000000000",
          irm: item.allocationMarket.irmAddress,
          lltv: item.allocationMarket.lltv.toString(),
        },
```